Use MCNP mesh tally data

bquan0 opened the pull request:
### Overview
Closes #2. This PR adds the `RadioactiveTank` scene, which represents a voxelized radiation field based on data generated from a MCNP mesh tally (courtesy of Tim Bohm). A short demo of the new features can be seen in the #svalinn-vr channel on Slack. 

### RadioactiveTank.gd
In `_ready()`, the script loads in the data from `tank10.csv` (not included in repo) into 3 arrays for the bounds of the voxels in the x, y, z directions. It also loads in the centroid (x,y,z), radiation value, and relative error of each voxel into another array.
In `get_radiation()`, the script calculates the index of the voxel that the player is in using the position of the player and the arrays of x, y, and z boundaries, and returns the radiation value of that voxel as well as the centroid of the voxel for display purposes. 

### New debug/visualization feature
For all radioactive objects (Tank and Rod), I added a **red mesh** that highlights the current voxel in the radiation field that the player is in. Currently, for the RadioactiveTank, the red mesh is much smaller than the player mesh, so the player mesh needs to be made invisible in order to see this red mesh. 

### MCNP data size
Something worth noting is that the game now takes around 1m 30s to load due to the large size of the MCNP data. Despite this, there doesn't seem to be any drops in performance when running the game (though I haven't tested it with multiple RadioactiveTanks). The size of the voxels for this MCNP data is 10x15x20cm, which might be too fine grain for the application of a VR game.

### Other notes:
Things not included in this PR:
- The `tank10.csv` file because it's around 200MB. 
- The python script used to generate the `tank10.csv` file from the original MCNP mesh tally `.txt` file because it's not used in the Godot game. Nevertheless, I am open to adding it here or in another repo if it's useful.

Other changes:
- Made the blue and red axes of the radioactive rod not "visible" as it can be confusing for people who pick up this 

As a final aside, the index calculation of the voxel for RadioactiveTank really threw me for a loop because I did it incorrectly for the RadioactiveRod (but it didn't matter there because all the max_indices were the same).

bquan0 committed:
add radioactive tank with MCNP mesh tally data

bquan0 committed:
clean up debug print statements in RadioactiveTank.gd

gonuke commented on scripts/RadioactiveTank.gd:
```diff
@@ -0,0 +1,136 @@
+	var x_index = 0
+	var rel_x = x - translation.x # position relative to this tank
+	while x_index < bounds_x.size() and rel_x >= bounds_x[x_index]:
+		x_index = x_index + 1
```
I recommend implementing a binary search in a function that you can reuse for each dimension.

It will also allow you to return a value of -1 if outside the grid in that dimension.  You can test this first before the search and return early. If so, you can also return early from this function with a null result. If the search returned -1 for out of range, you could just compare the product of the indices already set to 0.

Could also include the translation, or just apply it in the function call.

gonuke commented on scripts/RadioactiveTank.gd:
```diff
@@ -0,0 +1,136 @@
+	# get x boundaries
+	var csv_line = csv_file.get_csv_line()
+	for i in range(0, csv_line.size()):
+		bounds_x.append(float(csv_line[i]) / 100);
```
It's tempting to suggest a function for this that you can reuse with each dimension.

```
func _parse_dims(csv_line, scale)
    bounds = []
    for i in range(0,csv_line.size())
         bounds.append(csv_line[i] * scale)
    return bounds

scale = 1.0/100
bounds_x = _parse_dims(csv_file.get_csv_line(), scale)
bounds_y = _parse_dims(csv_file.get_csv_line(), scale)
bounds_z = _parse_dims(csv_file.get_csv_line(), scale)
```

gonuke commented on scripts/RadioactiveTank.gd:
```diff
@@ -0,0 +1,136 @@
+			$CurrentVoxel.visible = true
+	else:
+		# don't show the mesh if input position isn't a voxel in the map
+		$CurrentVoxel.visible = false
```
If you initialize this to false, you can return early from this function without having to worry about setting it.

gonuke commented on scripts/RadioactiveTank.gd:
```diff
@@ -0,0 +1,136 @@
+		centroid.y = radiation_map[index][1]
+		centroid.z = radiation_map[index][2]
+		
+		if index != prev_index:
```
Why not test this before testing if in range?

bquan0 replied:
Testing whether we should update the visualization mesh is only needed if the player is in the radiation field (index is in range)

gonuke replied:
Yep - I got that... in the previous structure, I thought this test could come even earlier, but it's moot after your recent updates.

gonuke commented on scripts/RadioactiveTank.gd:
```diff
@@ -0,0 +1,136 @@
+		for i in range(0, csv_line.size()):
+			# first 3 values are x,y,z of centroid in cm, so divide by 100
+			if i <= 2:
+				voxel.append(float(csv_line[i]) / 100)
```
Can a list in this language contain different types? Could it contain a `Vector3` object and also the 2 floats? You only use the centroid data to populate/modify a `Vector3` object, so it would make your later code simpler.

Also remove this magic number in favor of a scaling variable.

bquan0 replied:
Yep, it can. One example already in the code is the return value of the function, where I return an array with a float and a Vector3 (radiation value and centroid respectively). I will apply these changes.

gonuke reviewed (changes requested):
A few suggestions for performance and simplicity

bquan0 committed:
add binary search function, get_bounds function, and other optimizations

bquan0 committed:
revise comments, simplify mesh position setting

bquan0 commented:
- Added the `binary_search()` function
- Added the `get_bounds()` function (renamed from `parse_dims()` to fit the preexisting naming schemes)
- Added `scale` variable in `_ready()` to remove magic number of 100
- Made `radiation_map` store a `Vector3` of the centroids instead of 3 floats for the (x,y,z) positions
- Updated logic in `get_radiation()`, added optimizations now that `radiation_map` stores `Vector3`s

gonuke commented on scripts/RadioactiveTank.gd:
```diff
@@ -0,0 +1,149 @@
+		var csv_line = csv_file.get_csv_line()
+		var voxel = []
+		var voxel_centroid = Vector3.ZERO
+		for i in range(0, csv_line.size()):
+			# first 3 values are x,y,z of centroid in cm, so multiply by scale
+			if i == 0:
+				voxel_centroid.x = float(csv_line[i]) * scale
+			elif i == 1:
+				voxel_centroid.y = float(csv_line[i]) * scale
+			elif i == 2:
+				voxel_centroid.z = float(csv_line[i]) * scale
+				voxel.append(voxel_centroid)
+			# next 2 values are not in cm, so take them as is
+			else:
+				voxel.append(float(csv_line[i]))
```
```suggestion
		# first 3 values are x,y,z of centroid in cm, so multiply by scale
		voxel_centroid.x = float(csv_line[0]) * scale
		voxel_centroid.y = float(csv_line[1]) * scale
		voxel_centroid.z = float(csv_line[2]) * scale
		for i in range(3, csv_line.size()):
			voxel.append(float(csv_line[i]))
```

gonuke reviewed (changes requested):
one small simplification

bquan0 committed:
Add simplification for loading voxel data to radiation_map

Co-authored-by: Paul Wilson <paul.wilson@wisc.edu>

bquan0 committed:
add back append centroid, add if statement to prevent edge cases

bquan0 commented:
I tried adding the simplification, but ran into an `invalid index` error when loading the game. Must be an edge case when I get a csv line from the end of the file that doesn't have any voxel data (I think my generation python script puts a newline after all the voxel data lines). I fixed this by adding an if statement to check if the `csv_line` has 5 items.

gonuke reviewed (approved):
LGTM - Thanks @bquan0
